src/replication.c

```diff
@@ -4136,6 +4136,8 @@ void replicationCachePrimary(client *c) {
     serverAssert(server.primary != NULL && server.cached_primary == NULL);
     serverLog(LL_NOTICE, "Caching the disconnected primary state.");
 
+    /* Wait for IO operations to be done before proceeding */
```

Comment:
Why? I guess I don't immediately understand why we need to wait for IO on these client unlinks, but not the other places we free the client.

Reply:
I'm not sure if this is required, but I thought it would be safer to check that no I/O operations are running before unlinking the client.  After thinking about it, we already wait for client I/O in freeClient, so this check isn't needed here since it's called from freeClient. I moved this check inside unlinkClient since it might be called from other places without calling freeClient first.